plugins/arkworks/src/poseidon/test.rs

```diff
@@ -0,0 +1,472 @@
+// move tests that need arkworks types here
+
+#[cfg(feature = "ark-bls12-381")]
+use {
+    crate::{constraint::fp::Fp, ff::field_new, poseidon::NativeField},
+    ark_bls12_381::Fr,
+};
+
+#[cfg(feature = "ark-bls12-381")]
+mod round_constants {
+    use super::*;
+    use openzl_crypto::poseidon::round_constants::{generate_lfsr, sample_field_element};
+
+    #[allow(clippy::needless_borrow)] // NOTE: Clippy false positive https://github.com/rust-lang/rust-clippy/issues/9710
+    /// Checks if [`GrainLFSR`] matches hardcoded sage outputs.
+    #[test]
+    fn grain_lfsr_is_consistent() {
+        let test_cases = include!("parameters_hardcoded_test/lfsr_values");
+        let mut lfsr = generate_lfsr(255, 3, 8, 55); // TODO: Change for Bn254
+        for x in test_cases {
+            assert_eq!(sample_field_element::<Fp<Fr>, _>(&mut lfsr), x);
+        }
+    }
+}
+
+#[cfg(feature = "ark-bls12-381")]
+mod matrix {
+    use super::*;
+    use openzl_crypto::poseidon::matrix::{
+        inner_product, vec_add, Matrix, MatrixOperations, SquareMatrix,
+    };
+
+    /// Checks if generating minor matrix is correct.
+    #[test]
+    fn minor_is_correct() {
```

Comment:
This belongs in `matrix`, or at the very least, the abstract part of it can be factored out there.

Reply:
Yes that should be done whenever we move linear algebra out of poseidon #41